### PR TITLE
Include animation/transition durations

### PR DESCRIPTION
**Pull request by beefchimi**
> **Note:** First time contributor - not quite sure what steps I need to perform. I am definitely missing a `docs` entry. Also not sure if I need a dedicated `duration` task in the gulpfile. Any advice is appreciated.

This PR adds all the values for [duration from Polaris](https://github.com/Shopify/polaris-react/blob/master/src/styles/foundation/duration.scss#L1-L9).

We need to utilize some of these `duration` values to pass along to `react-transition-group`. Currently, we have needed to manually copy the equivalent values from SCSS into our JS. This PR aims to resolve that by allowing us to instead import the `duration` values, preventing manual updates in the future _(if the SCSS values were to ever change)_.

### CODE REVIEW

**Comment by alex-page** [comment]
@beefchimi Thanks for the contribution, the `polaris-tokens` team will be in touch this week. I had a quick skim over the files changed and it looks good to me!

File: tokens/duration.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  - name: duration-fast
+    value: 100
+  - name: duration-base
+    value: 200
+  - name: duration-slow
+    value: 300
+  - name: duration-slower
+    value: 400
+  - name: duration-slowest
+    value: 500
```
Comment (kaelig) [inline tokens/duration.yml]: Is there a reason those values are unitless?

Comment (johanstromqvist) [reply tokens/duration.yml]: Thanks for ping!

- Whatever solves your problems now is great
- My main focus will be guidelines of usage, rather than spec of tokens
- 50ms increments will be very useful
- Current naming convention looks perfect to me.
- More elaborate token naming is hinged on more elaborate token concepts. We currently have no such concepts or time allocated to create such concepts. An exploration of this would quite naturally follow when we have a couple of UX patterns that we'd want to streamline.

Can't respond to the `units` question, I assume that it's not meant for me.

Comment (beefchimi) [reply tokens/duration.yml]: @kaelig right, so - in JS land we want a unitless value, since JS understands that it is milliseconds. In SCSS land we want this as a string with `ms`.

What is the correct solution for this? Do we:
1. use string interpolation when consumed in SCSS to append the `ms`, or
2. write a custom task to handle `duration`, sending the unchanged `number` to everything expect SCSS, where we need to append `ms` and return a `string

Comment (kaelig) [reply tokens/duration.yml]: Thanks @johanstromqvist! 

@beefchimi I looked into it and I think the best way to think about it is to have units in the source tokens (more explicit, could even be authored as seconds or hours), and strip them when appropriate (also: convert them to milliseconds, because in JS that's most useful).

I've opened a PR that should solve this: #55

**Comment by beefchimi** [comment]
@sarahill I've opened an issue on `polaris-react` to track the upgrade:
https://github.com/Shopify/polaris-react/issues/1258

@kaelig Should I me committing the results of running `dev dist` in this PR? Is there anything else we want to change with the code before doing that?

**Comment by kaelig** [comment]
Hey, sorry for the late response!

Here's a Shipit Stack for deploying from this PR's branch - https://shipit.shopify.io/shopify/polaris-tokens/next

I just published `v2.4.0-beta.1`, let me know how it goes for you.

**Comment by beefchimi** [comment]
@kaelig opened a PR on `polaris-react`: https://github.com/Shopify/polaris-react/pull/1268

Did we forget to build a `duration.duration-map.scss`?

**Comment by kaelig** [comment]
It depends, what did you expect to be in that file?

Does https://github.com/Shopify/polaris-tokens/blob/duration/dist/duration.map.scss work for you?

**Comment by beefchimi** [comment]
@kaelig that file works, I guess im just curious about what Polaris prefers. More context is given in the PR linked above. If you get the chance later today or next week, I'd be interested to get your feedback.

File: __tests__/duration.test.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    expect(
+      _(durationJSON)
+        .toPairs()
+        .sortBy(0)
+        .fromPairs()
+        .value(),
+    ).toMatchSnapshot();
+  });
+
+  it('outputs integers in JSON', () => {
```
Comment (beefchimi) [inline __tests__/duration.test.js]: Not sure how useful this test actually is... would probably be better to loop through each key in the object and assert that each is a `number`. Couldn't find any `jest` methods for this - wasn't sure of the best way to author that assertion.

Comment (kaelig) [reply __tests__/duration.test.js]: Would an asserrstion on Object.values + every work?

**Comment by kaelig** [comment]
For consistency, we might want to create a duration.duration-map.scss file. However, this type of custom map is becoming extra work for each format that we add... So we could also decide to break that cycle and simply start using the existing map, resulting in less work in the future. Thoughts?

File: __tests__/duration.test.js (hunk 2 above)
Comment (beefchimi) [reply __tests__/duration.test.js]: Brilliant! This is what I did:

```
  it('outputs integers in JSON', () => {
    const everyValueIsInteger = Object.values(durationJSON).every((value) =>
      Number.isInteger(value),
    );
    expect(everyValueIsInteger).toBe(true);
  });
```

**Comment by beefchimi** [comment]
@kaelig 

>  this type of custom map is becoming extra work for each format that we add...

Yea I feel the same.

### Just to clarify the current state of things:

**By default _(as far as SCSS is concerned)_, we output:**
1. A file containing `variables` for each value.
2. A file containing a `map`, with each `key` mirroring the formers `variable` name, and each `value` being a SCSS `list` type _(because our values could be nested maps)_

**Example: spacing.scss**
```scss
$spacing-none: 0;
$spacing-extra-tight: 4px;
$spacing-tight: 8px;
$spacing-base-tight: 12px;
$spacing-base: 16px;
$spacing-loose: 20px;
$spacing-extra-loose: 32px;
```

**Example: spacing.map.scss**
```scss
$polaris-spacing-map: (
  'spacing-none': (
    0,
  ),
  'spacing-extra-tight': (
    4px,
  ),
  'spacing-tight': (
    8px,
  ),
  'spacing-base-tight': (
    12px,
  ),
  'spacing-base': (
    16px,
  ),
  'spacing-loose': (
    20px,
  ),
  'spacing-extra-loose': (
    32px,
  ),
);
```

At some point, we crafted a "special" map for `polaris` consumption, because thats how the original `map` + corresponding `function` was authored:

**Example: spacing.spacing-map.scss**
```scss
$polaris-spacing: (
  'none': 0,
  'extra-tight': 4px,
  'tight': 8px,
  'base-tight': 12px,
  'base': 16px,
  'loose': 20px,
  'extra-loose': 32px,
);
```

**Example: Polaris spacing() function**
```scss
@function spacing($variant: base) {
  $fetched-value: map-get($spacing-data, $variant);

  @if type-of($fetched-value) == number {
    @return rem($fetched-value);
  } @else {
    @error 'Spacing variant `#{$variant}` not found. Available variants: #{available-names($spacing-data)}';
  }
}
```

I think we either do one or the other. So we either go all in on what Polaris has traditionally expected _(spacing.spacing-map.scss)_, or stick with our more "verbose" map type `(spacing.map.scss)`.

Since it looks like `polaris-tokens` currently authors _more code_ to accomodate for the "traditional polaris style" _(spacing.spacing-map.scss)_, I'm inclined to say **let's break the mold** and remove these special maps, going all in on the "verbose" style.

The change required in Polaris is trivial. The `spacing` function referenced above changes to look like:

**Example: New Polaris function**
```scss
@function spacing($variant: base) {
  $interpolated-value: 'spacing-' + $variant;
  $fetched-value: nth(map-get($spacing-data, $interpolated-value), 1);

  @if type-of($fetched-value) == number {
    @return rem($fetched-value);
  } @else {
    @error 'Spacing variant `#{$interpolated-value}` not found. Available variants: #{available-names($spacing-data)}';
  }
}
```

^ Thats a `3` line change. Nothing significant. Just my 2 cents, but seeing as though no one has weighed in on this change in my [polaris-react PR](https://github.com/Shopify/polaris-react/pull/1268), I'd say we can make the decision.

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
```
Comment (kaelig) [inline CHANGELOG.md]:
```suggestion
<!-- ## [Unreleased] -->
```

File: gulpfile.js [hunk 4]
```diff
@@ -225,6 +235,7 @@ gulp.task(
     'web-formats',
     'typings',
     'color-filters',
+
```
Comment (kaelig) [inline gulpfile.js]: Can we remove this line?

File: gulpfile.js [hunk 5]
```diff
@@ -11,8 +12,17 @@ theo.registerValueTransform(
   (prop) => prop.get('type') === 'color',
   (prop) => prop.getIn(['meta', 'filter']) || 'none',
 );
+
 theo.registerTransform('filter', ['filter']);
 
+theo.registerValueTransform(
+  'timing/ms-unitless',
+  (prop) => prop.get('type') === 'time',
+  (prop) => (prop.get('value') === 0 ? 0 : ms(prop.get('value'))),
```
Comment (kaelig) [inline gulpfile.js]: I just realized we also need a test that ensures that in `duration.raw.json`, `duration-none`'s value should be unitless, while the other values have `ms` in them.

File: package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@shopify/polaris-tokens",
-  "version": "2.4.0",
+  "version": "2.5.0",
```
Comment (kaelig) [inline package.json]: You can revert this change, as the version number will be bumped by `yarn version` (which we perform on master directly, and may require an admin to do it).

File: gulpfile.js (hunk 5 above)
Comment (beefchimi) [reply gulpfile.js]: Incoming new commit! Please take a look and let me know if I did it correctly :)

Comment (kaelig) [reply gulpfile.js]: Looks good! 👍

Comment (kaelig) [reply gulpfile.js]: Looks good! 👍
